src/settings/sync_with_cards.c

```diff
@@ -0,0 +1,323 @@
+
+/**
+ * @brief Helper function to store the wallet data in the X1 vault flash memory.
+ *
+ * @param wallet_list The list of wallets present in an X1 card
+ * @param tapped_card The encoded value of the tapped card
+ */
+static void sync_wallets_in_flash(const wallet_list_t *wallet_list,
+                                  const uint8_t *tapped_card);
+
+/**
+ * @brief Helper function to check if X1 Vault can be synced with the cards.
+ * @details Checks if atleast 2 cards are paired and no wallet is in partial,
+ * invalid or locked state. If the device state is not eligible for syncing an
+ * appropriate error message is set
+ *
+ * @return true If the device state is eligible for syncing
+ * @return false If the device state is not eligible for syncing
+ */
+bool sync_with_card_eligibility(void);
+
+/*****************************************************************************
+ * STATIC VARIABLES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * GLOBAL VARIABLES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * STATIC FUNCTIONS
+ *****************************************************************************/
+static bool get_wallets_from_card(wallet_list_t *wallet_list,
+                                  uint8_t *tapped_card) {
+  card_fetch_wallet_list_config_t configuration = {0};
+  card_fetch_wallet_list_response_t response = {0};
+
+  configuration.acceptable_cards = ACCEPTABLE_CARDS_ALL;
+  configuration.heading = ui_text_tap_1_2_cards;
+  configuration.msg = ui_text_place_card_below;
+  configuration.skip_card_removal = false;
+
+  response.wallet_list = wallet_list;
+  response.tapped_card = 0;
+
+  if (!card_fetch_wallet_list(&configuration, &response)) {
+    return false;
+  }
+
+  *tapped_card = response.tapped_card;
+  return true;
+}
+
+static void sync_wallets_in_flash(const wallet_list_t *wallet_list,
+                                  const uint8_t *tapped_card) {
+  for (uint8_t index = 0; index < wallet_list->count; index++) {
+    Flash_Wallet new_wallet = {0};
+
+    new_wallet.state = VALID_WALLET_WITHOUT_DEVICE_SHARE;
+    new_wallet.wallet_info = wallet_list->wallet[index].info;
+    new_wallet.is_wallet_locked = wallet_list->wallet[index].locked;
+    // assuming if person is syncing then all cards have share
+    new_wallet.cards_states = 15;
+
+    memcpy(new_wallet.wallet_name, wallet_list->wallet[index].name, NAME_SIZE);
+    memcpy(new_wallet.wallet_id, wallet_list->wallet[index].id, WALLET_ID_SIZE);
+
+    // Record card number on which the wallet is locked
+    if (new_wallet.is_wallet_locked) {
+      new_wallet.challenge.card_locked = *tapped_card;
+    }
+
+    uint32_t dummy;
+    add_wallet_to_flash(&new_wallet, &dummy);
```

Comment:
Please check the return code here. Either assert or return the error code.

Reply:
Ok, sure

Reply:
Added the handling